x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDatafeedsStatsAction.java

```diff
@@ -58,22 +63,37 @@ protected void masterOperation(GetDatafeedsStatsAction.Request request, ClusterS
                                    ActionListener<GetDatafeedsStatsAction.Response> listener) throws Exception {
         logger.debug("Get stats for datafeed '{}'", request.getDatafeedId());
 
-        datafeedConfigProvider.expandDatafeedIds(request.getDatafeedId(), request.allowNoDatafeeds(), ActionListener.wrap(
-                expandedDatafeedIds -> {
-                    PersistentTasksCustomMetaData tasksInProgress = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
-                    List<GetDatafeedsStatsAction.Response.DatafeedStats> results = expandedDatafeedIds.stream()
-                            .map(datafeedId -> getDatafeedStats(datafeedId, state, tasksInProgress))
-                            .collect(Collectors.toList());
-                    QueryPage<GetDatafeedsStatsAction.Response.DatafeedStats> statsPage = new QueryPage<>(results, results.size(),
-                            DatafeedConfig.RESULTS_FIELD);
-                    listener.onResponse(new GetDatafeedsStatsAction.Response(statsPage));
+        datafeedConfigProvider.expandDatafeedConfigs(
+            request.getDatafeedId(),
+            request.allowNoDatafeeds(),
+            ActionListener.wrap(
+                datafeedBuilders -> {
+                    Map<String, String> jobIdByDatafeedId = datafeedBuilders.stream()
+                        .map(DatafeedConfig.Builder::build)
+                        .collect(Collectors.toUnmodifiableMap(DatafeedConfig::getId, DatafeedConfig::getJobId));
+                    jobResultsProvider.datafeedTimingStats(
+                        jobIdByDatafeedId.values(),
+                        timingStatsByJobId -> {
+                            PersistentTasksCustomMetaData tasksInProgress = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+                            List<GetDatafeedsStatsAction.Response.DatafeedStats> results = datafeedBuilders.stream()
+                                .map(datafeedBuilder -> getDatafeedStats(
+                                    datafeedBuilder.getId(), state, tasksInProgress, jobIdByDatafeedId, timingStatsByJobId))
```

Comment:
```suggestion
                                    datafeedBuilder.getId(), state, tasksInProgress, jobIdByDatafeedId.get(datafeedBuilder.getId()), timingStatsByJobId))
```
you should just pass `jobId` and not the whole map.